src/sympc/protocol/aby3/aby3.py

```diff
@@ -169,3 +207,77 @@ def _get_truncation_pair(x: MPCTensor, session: Session) -> Tuple[MPCTensor]:
         r_mpc = MPCTensor(shares=r, session=session, shape=x.shape)
         rPrime_mpc = MPCTensor(shares=rPrime, session=session, shape=x.shape)
         return r_mpc, rPrime_mpc
+
+    @staticmethod
+    def local_decomposition(
+        x: ReplicatedSharedTensor, ring_size: str
+    ) -> List[ReplicatedSharedTensor]:
+        """Performs local decomposition to generate shares of shares.
+
+        Args:
+            x (ReplicatedSharedTensor) : input RSTensor.
+            ring_size (str) : Ring size to generate decomposed shares in.
+
+        Returns:
+            List[ReplicatedSharedTensor]: Decomposed shares in the given ring size.
+
+        Raises:
+            ValueError: If RSTensor does not have session uuid.
+        """
+        if x.session_uuid is None:
+            raise ValueError("Input RSTensor should have session_uuid")
+
+        session = get_session(x.session_uuid)
+        ring_size = int(ring_size)
+        tensor_type = get_type_from_ring(ring_size)
+        rank = session.rank
+        nr_parties = session.nr_parties
+
+        zero = torch.zeros(x.shares[0].shape).type(tensor_type)
+
+        shares = [[zero.clone(), zero.clone()] for i in range(nr_parties)]
+
+        shares[rank][0] = x.shares[0].clone().type(tensor_type)
+
+        shares[(rank + 1) % nr_parties][1] = x.shares[1].clone().type(tensor_type)
+
+        rst_list = []
+        for i in range(nr_parties):
+            rst = x.clone()
+            rst.shares = shares[i]
+            rst.ring_size = ring_size
+            rst_list.append(rst)
+
+        return rst_list
+
+    @staticmethod
+    def bit_injection(x: MPCTensor, session: Session, ring_size: int) -> MPCTensor:
+        """Perfom ABY3 bit injection for conversion of binary share to arithmetic share.
+
+        Args:
+            x (MPCTensor) : MPCTensor with shares of bit.
+            session (Session): session the share belongs to.
+            ring_size (int) : Ring size of arithmetic share to convert.
+
+        Returns:
+            arith_share (MPCTensor): Arithmetic shares of bit in input ring size.
+        """
+        args = [[share, str(ring_size)] for share in x.share_ptrs]
+
+        decompose = parallel_execution(ABY3.local_decomposition, session.parties)(args)
+
+        x1_sh, x2_sh, x3_sh = list(zip(*decompose))
+
+        x1_sh = list(map(lambda pt: pt.resolve_pointer_type(), x1_sh))
```

Comment:
MAYBE: Better use list comprehension here:
[pt.resolve_pointer_type() for pr in x1_sh]

Reply:
Good suggestion , list comprehension would make it simple :+1: .